code/modules/wiremod/shell/brain_computer_interface.dm

```diff
@@ -0,0 +1,544 @@
+	var/obj/item/organ/cyberimp/bci/bci = shell
+	if (!istype(bci))
+		CRASH("BCI action button was placed inside [shell] ([shell.type]), not a BCI")
+
+	bci_action = new(src)
+	update_action()
+
+	bci.actions += list(bci_action)
+
+/obj/item/circuit_component/bci_action/unregister_shell(atom/movable/shell)
+	var/obj/item/organ/cyberimp/bci/bci = shell
+	if (!istype(bci))
+		CRASH("BCI action button was unregistered for [shell] ([shell.type]), not a BCI")
+
+	bci.actions -= bci_action
+	QDEL_NULL(bci_action)
+
+/obj/item/circuit_component/bci_action/input_received(datum/port/input/port)
+	. = ..()
+
+	if (.)
+		return
+
+	if (!isnull(bci_action))
+		update_action()
+
+/obj/item/circuit_component/bci_action/proc/update_action()
+	bci_action.name = button_name.input_value
+	bci_action.button_icon_state = "nanite_[replacetextEx(lowertext(current_option), " ", "_")]"
+
+/datum/action/innate/bci_action
+	name = "Action"
+	icon_icon = 'icons/mob/actions/actions_items.dmi'
+	check_flags = AB_CHECK_CONSCIOUS
+	button_icon_state = "nanite_power"
+
+	var/obj/item/circuit_component/bci_action/circuit_component
+
+/datum/action/innate/bci_action/New(obj/item/circuit_component/bci_action/circuit_component)
+	..()
+
+	src.circuit_component = circuit_component
+
+/datum/action/innate/bci_action/Destroy()
+	circuit_component.bci_action = null
+	circuit_component = null
+
+	return ..()
+
+/datum/action/innate/bci_action/Activate()
+	circuit_component.signal.set_output(COMPONENT_SIGNAL)
+
+/obj/item/circuit_component/bci_core
+	display_name = "BCI Core"
+	display_desc = "Controls the core operations of the BCI."
+
+	/// A reference to the action button to look at charge/get info
+	var/datum/action/innate/bci_charge_action/charge_action
+
+	var/datum/port/input/message
+	var/datum/port/input/send_message_signal
+
+	var/datum/port/output/user_port
+
+	var/datum/weakref/user
+
+/obj/item/circuit_component/bci_core/Initialize()
+	. = ..()
+
+	message = add_input_port("Message", PORT_TYPE_STRING)
+	send_message_signal = add_input_port("Send Message", PORT_TYPE_SIGNAL)
+
+	user_port = add_output_port("User", PORT_TYPE_ATOM)
+
+/obj/item/circuit_component/bci_core/Destroy()
+	QDEL_NULL(charge_action)
+
+	message = null
+	send_message_signal = null
+
+	user_port = null
+
+	user = null
+
+	return ..()
+
+/obj/item/circuit_component/bci_core/register_shell(atom/movable/shell)
+	var/obj/item/organ/cyberimp/bci/bci = shell
+	if (!istype(bci))
+		CRASH("BCI charge listener was placed inside [shell] ([shell.type]), not a BCI")
+
+	charge_action = new(src)
+	bci.actions += list(charge_action)
+
+	RegisterSignal(shell, COMSIG_ORGAN_IMPLANTED, .proc/on_organ_implanted)
+	RegisterSignal(shell, COMSIG_ORGAN_REMOVED, .proc/on_organ_removed)
+
+/obj/item/circuit_component/bci_core/unregister_shell(atom/movable/shell)
+	var/obj/item/organ/cyberimp/bci/bci = shell
+	if (!istype(bci))
+		CRASH("BCI charge listener was unregistered for [shell] ([shell.type]), not a BCI")
```

Comment:
I don't think you should CRASH here if other types can be supported.
The good thing about making the behaviour as generic as possible is that it means it can easily be implemented in other shells, or for example, implemented in an admin-made shell (e.g. a different organ).

Reply:
Ehh, probably fair.